File: packages/database/src/schema/sessionSeries.ts
import { relations } from "drizzle-orm";
import {
	foreignKey,
	index,
	integer,
	pgTable,
	time,
	timestamp,
	uuid,
	varchar,
} from "drizzle-orm/pg-core";
import { campaignsTable } from "./campaigns";
import { sessionExceptionsTable } from "./sessionExceptions";
import { sessionsTable } from "./sessions";

export const sessionSeriesTable = pgTable(
	"session_series",
	{
		campaignId: uuid("campaign_id").notNull(),
		createdAt: timestamp("created_at", { mode: "date" }).defaultNow().notNull(),
		description: varchar("description"),
		discordEventId: varchar("discord_event_id"),
		durationMinutes: integer("duration_minutes").notNull().default(180),
		googleCalendarEventId: varchar("google_calendar_event_id"),
		id: uuid("id").primaryKey().defaultRandom(),
		pollId: varchar("poll_id"),
		rrule: varchar("rrule"),
		seriesEndDate: timestamp("series_end_date", { mode: "date" }),
		seriesStartDate: timestamp("series_start_date", { mode: "date" }).notNull(),
		startTime: time("start_time"),
		timezone: varchar("timezone").notNull().default("UTC"),
		title: varchar("title").notNull(),
		updatedAt: timestamp("updated_at", { mode: "date" })
			.defaultNow()
			.notNull()
			.$onUpdate(() => new Date()),
	},
	(t) => [
		foreignKey({
			columns: [t.campaignId],
			foreignColumns: [campaignsTable.id],
			name: "fk_session_series_campaign_id",
		}).onDelete("cascade"),
		index("idx_session_series_campaign_id").on(t.campaignId),
		index("idx_session_series_discord_event_id").on(t.discordEventId),
	],
);

export const sessionSeriesRelations = relations(
	sessionSeriesTable,
	({ one, many }) => ({
		campaign: one(campaignsTable, {
			fields: [sessionSeriesTable.campaignId],
			references: [campaignsTable.id],
		}),
		exceptions: many(sessionExceptionsTable),
		sessions: many(sessionsTable),
	}),
);

File: apps/api/internal/db/session_series.go
package db

import (
	model "github.com/BBruington/party-planner/api/internal/models"
)

const sessionSeriesColumns = `id, campaign_id, title, description, discord_event_id, google_calendar_event_id, poll_id, rrule, start_time, series_start_date, series_end_date, created_at, updated_at, timezone, duration_minutes`

func scanSessionSeries(row interface{ Scan(...any) error }) (*model.SessionSeries, error) {
	var s model.SessionSeries
	err := row.Scan(
		&s.ID, &s.CampaignID, &s.Title, &s.Description,
		&s.DiscordEventID, &s.GoogleCalendarEventID, &s.PollID,
		&s.RRule, &s.StartTime,
		&s.SeriesStartDate, &s.SeriesEndDate, &s.CreatedAt, &s.UpdatedAt, &s.Timezone, &s.DurationMinutes,
	)
	if err != nil {
		return nil, err
	}
	return &s, nil
}

func (db *DB) GetSessionSeriesByDiscordEventID(discordEventID string) (*model.SessionSeries, error) {
	row := db.conn.QueryRow(`SELECT `+sessionSeriesColumns+` FROM session_series WHERE discord_event_id = $1 LIMIT 1`, discordEventID)
	return scanSessionSeries(row)
}
